### ISSUE TITLE
The Why of the Juvix Macro System

### ISSUE THREAD

**Issue by mariari** [heliaxdev/juvix#748]
## The Why of the Juvix Macro System

Juvix is a high level functional programming language herring from the proud lineage of the ML family. While the ML family is a wonderful lineage for exploring type theory and various mathematical abstractions, it has long diverged from the lineage in which it was originally inspired from, LISP. What this means is that while ML was advancing type theory, it fixed the core environment from in which it was interacted with in order to have better fixed properties. Thus we drop talking about the living changing environment that our code lives in, we don't design our languages with being able to abstract over it in mind. In time these features are often added back see [1][2][3], some of these like [1] take the type system as a macro system and play off these ideas, while fixing the forms into the compiler while [2] and [3] allow new forms but are either cumbersome to use and may break between versions.

This would all be fine if Juvix just wanted to further type theory, however our goals are much broader than that. We want to make a language that helps us write code for various domains like, smart contracts, compilers, theorem provers, proof assistants, Graphical appliations, etc. And in order to do that, we must care about the changing world. Thus we are image based (read #742 for more details), however that isn't enough to satisfy the demands of the various domains we wish to handle.

#### A small Example

Let us consider a motivating example, first in a version of Juvix without macros

NOTE :: we use ocaml style comments just to have better syntax highlighting on github
```ocaml

sig my-key : string
sig my-account : account
sig key : account
sig storage : type

sig accept-withdraws-from : account -> storage -> storage -> bool
let accept-withdraws-from key initial-storage final-storage =
  (* get the transaction that is occuring on this account *)
  (* let's say this is simplified and it's not a Maybe value *)
  let transaction = transaction-on my-account final-storage in
  let key-to      = Deserialize.account (lookup "to" transaction) in
  let my-pub-key  = Deserialize.account (lookup "account" transaction) in
  case (my-pub-key, key-to) of
    | (Just my-pub-key, Just key-to) =
      if | my-pub-key == my-account && key-to == key =
          let difference =
            let deseralize-amount x = account.amoutn
            let amount-start = account.amount intial-storage in
            let amount-end   = account.amount final-storage in
            amount-start - amount-end
          in difference < 10
   | _ =
    false -- don't accept transactions from any other key
```

This code is kind of unweidly, as we have to mix our concerns of deseralizing into the code. This of course can be somewhat abstracted away, we could make functions which take away part of the concern

```ocaml

type storage = {
  transaction : transaction-type,
}

type transaction-type = {
  account-from : account,
  account-to   : account
}

let to-storage-type = ... deseralize code here to fill the type>

sig accept-withdraws-from : account -> Chain.storage -> Chain.storage -> bool
let accept-withdraws-from key initial-storage final-storage =
  let storage = to-storage-type final-storage in
  if | storage.transaction.to == key =
      let difference =
         let starting = storage.account-from.amount in
         let ending  = storage.account-to.amount in
         starting - ending
     in difference - 10
    | else = false
```

This is better, and getting close to less boilerplate, but we had to derive our own deserializer via `to-storage-type`, what if the type of contract gets more verbose, or what if this changes and we want to deal with a different kind of storage? Well then we'd have to write new deseralizers from scratch.

However, we could from a template tell it what form to fill, our type `stroage`, and `transaction-type` mimiced fairly closely to what the names in storage are. So we could derrive a manual declassifer as follows

```ocaml
sig to-any-storage : symbol -> Sexp.list symbol -> Chain.storage -> Maybe ???
let to-any-storage typ nested-record-symbols contract =
  let lookup symb = lookup (to-string symb)  in
  let rec (Atom x : xs) acc =
    case lookup x of
    | Just res ->  rec xs (Record.insert x res)
    | Nothing -> Nothing
  let rec (x : xs) = rec xs (rec x Record.new)
  let rec  Nil acc = acc
  in rec nested-record-symbols
```

Hmm, this version has a few flaws, first is that it's a pain to always send in a literal nested list of names to properly lookup (maybe we can use the context for that, given the type name!?), and two we just get a naked record back that is rather hard to type other than the assertion of the symbol we have, but that seems unidiomatic.

Let us see how we'd want to write this contract if we could, let us forget the constraints of juvix

```ocaml

type storage = {
  transaction : transaction-type,
}

type transaction-type = {
  account-from : account,
  account-to   : account
}

sig-pred accept-withdraws-from : key -> storage -> storage -> bool
let-pred accept-withdraws-from key {transaction = initial} {transaction = final}
  | key == final.account-to =
    let difference = initial.account-from.balance - final.account-from.balance in
    difference < 10
```

This seems a lot more reasonable, in fact if we often deal with a particular key and want to build around that, we can further strip this form.

And even building a macro around this doesn't seem too daunting, in fact let us finish our to-any-storage from before

```ocaml
macro to-any-storage typ contract =
  record-form <- Context.lookup-type-record typ
  let nested-record-symbol = Record.from-sexp
  let lookup symb = lookup (to-string symb)  in
  let rec (Atom x : xs) acc =
    case lookup x of
    | Just res ->  rec xs (Record.insert x res)
    | Nothing -> Nothing in
  let rec (x : xs) = rec xs (rec x Record.new) in
  let rec  Nil acc = acc in
  let construct-name = Symbol.splice typ '-from-stroage 
  '(progn
      (sig ~construct-name : Chain.storage -> ~typ)
      (* this has been left to the reader as an exercise *)
      (let ~construct-name = ... )
rec nested-record-symbols
```

I didn't fill out the function definition, however it would be easy to see what we need to generate that.

After we make this we can combine it with sig-pred to generate the constructor for the function if it doesn't already exist (we can extend our program to store it in a centralized module that we can check to avoid duplication of code!)

The moral of the story is that macros allow us to extend the language into the domain of the problem when it does not neatly fit. While Juvix was designed for smart contracts, there will be countless domains that we could not have seen juvix being used. In these domains they will come to similar discoveries where the lanugage needs to extend but can't. Thus the Juvix Macro system is there to ensure that they can proper extend the language to fit their needs much like how projects like [4] and [5] extended lisp to properly model their domain.

### Issues To Track to get Macros Implemented 
- #671 

### References

[1] https://fstar-lang.org/papers/metafstar/
  - This uses the effect system as an extension with macros baked into the compiler to make it ergonomic
[2] https://wiki.haskell.org/Template_Haskell
    - added after haskell 98
[3] https://discuss.ocaml.org/t/the-future-of-ppx/3766
  - PPX is often so painful, as the underlying form changes every update!
      - "The main reason why ppx rewriters are so much pain is because the system is based on the compiler libraries. The compiler libraries are meant for experts and provide no stability guarantee. With such an unstable basis, it is no surprise that the whole system keeps breaking all the time."

[4] https://nyxt.atlas.engineer/
[5] https://github.com/cbaggers/cepl

**Comment by cwgoes** [comment]
Thank you for this write-up, particularly the motivating examples with validity predicates, very helpful.

One thing to keep in mind is that we care a lot about computational efficiency in deserialisation - in particular, it would be nice to have a kind of lazy deserialisation which allowed specification of the full type of the storage but didn't actually deserialise particular fields until they were accessed - do you think this sort of behaviour would be possible with macros? 

Another, related question - are macros always zero-runtime-cost? Can we ensure this e.g. with QTT or something?

**Comment by mariari** [comment]
> Thank you for this write-up, particularly the motivating examples with validity predicates, very helpful.
> 
> One thing to keep in mind is that we care a lot about computational efficiency in deserialisation - in particular, it would be nice to have a kind of lazy deserialisation which allowed specification of the full type of the storage but didn't actually deserialise particular fields until they were accessed - do you think this sort of behaviour would be possible with macros?

Yeah very possible, as we are generating code based on the type, what we can do is allow the automation of the same type but with lazy fields attached where we only pay the cost once we access them, the cool thing about generating code is that we can take the route that is most efficient without worrying about the interface!!!

> 
> Another, related question - are macros always zero-runtime-cost? Can we ensure this e.g. with QTT or something?

Macros are typically compile time cost wise, as they do generation of the syntax given to them.

However with that said, people have used macros at run time to generate efficient code for specific use cases as data comes in, and that costs cycles, but I'm not sure we'll use those for the block chain right away (we actually could, as if we have a type contract that pays to generate specalized code as it'll get called a lot that could be a more effectively rewiring system than a normal cache as we can have specalized logic based on use case of real world data! Really the model can work well esp if we can combine it with other systems).

**Comment by rokopt** [comment]
This issue might be a good place to note something we discussed this morning:  one thing we'll have to figure out is how macros/metaprogramming will interact with usages.

**Comment by rokopt** [comment]
> This issue might be a good place to note something we discussed this morning: one thing we'll have to figure out is how macros/metaprogramming will interact with usages.

(I just noticed a mention of this in https://github.com/anomanetwork/juvix/issues/256#issue-538439941 .)
